java/src/main/java/cucumber/runtime/java/ObjectFactoryLoader.java

```diff
@@ -53,6 +55,46 @@ public static ObjectFactory loadObjectFactory(ClassFinder classFinder, String ob
         return objectFactory;
     }
 
+    private static io.cucumber.core.backend.ObjectFactory loadSingleObjectFactory(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader) {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> objectFactories = loadObjectFactories(reflections, loader);
+        final io.cucumber.core.backend.ObjectFactory result;
+        if (objectFactories.isEmpty()) {
+            throw new NoInstancesException(io.cucumber.core.backend.ObjectFactory.class);
+        } else if (objectFactories.size() > 1) {
+            throw new TooManyInstancesException(objectFactories.values());
+        } else {
+            result = objectFactories.values().iterator().next();
+        }
+        return result;
+    }
+
+    private static Map<String, io.cucumber.core.backend.ObjectFactory> loadObjectFactories(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader) {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> knowObjectFactories = new HashMap<>();
+        final List<URI> packages = asList(URI.create("classpath:cucumber/runtime"));
+        loadObjectFactories(reflections.instantiateSubclasses(io.cucumber.core.backend.ObjectFactory.class, packages, new Class[0], new Object[0]), knowObjectFactories);
+        loadObjectFactories(loader, knowObjectFactories);
+        return knowObjectFactories;
+    }
+
+    private static io.cucumber.core.backend.ObjectFactory loadSelectedObjectFactory(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader, final ClassFinder classFinder, final String objectFactoryClassName) throws ClassNotFoundException {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> objectFactoriesByServiceLoader = new HashMap<>();
+        loadObjectFactories(loader, objectFactoriesByServiceLoader);
+        final io.cucumber.core.backend.ObjectFactory result;
+        if (objectFactoriesByServiceLoader.containsKey(objectFactoryClassName)) {
+            result = objectFactoriesByServiceLoader.get(objectFactoryClassName);
+        } else {
+            final Class<io.cucumber.core.backend.ObjectFactory> objectFactoryClass = (Class<io.cucumber.core.backend.ObjectFactory>) classFinder.loadClass(objectFactoryClassName);
+            result = reflections.newInstance(new Class[0], new Object[0], objectFactoryClass);
+        }
+        return result;
```

Comment:
It's okay to use an early return statement once an object factory has been found.